sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md

```diff
@@ -0,0 +1,1347 @@
+// service bus client above, this would be where application methods sending a batch would be
+// called.
+
+ServiceBusSender sender = client.CreateSender("mockQueueName");
+
+ServiceBusMessageBatch batch = await sender.CreateMessageBatchAsync(CancellationToken.None);
+
+// This is creating a list of messages to use in our test.
+
+List<ServiceBusMessage> sourceMessages = new();
+
+for (int index = 0; index < batchCountThreshold; index++)
+{
+    var message = new ServiceBusMessage($"Sample-Message-{index}");
+    sourceMessages.Add(message);
+}
+
+// Here we are adding messages to the batch. They should all be accepted.
+
+foreach (var message in sourceMessages)
+{
+    Assert.True(batch.TryAddMessage(message));
+}
+
+// Since there are already batchCountThreshold number of messages in the batch,
+// this message will be rejected from the batch.
+
+Assert.IsFalse(batch.TryAddMessage(new ServiceBusMessage("Too Many Events.")));
+
+// For illustrative purposes we are calling SendMessagesAsync. Application-defined methods
+// would be called here instead.
+
+await sender.SendMessagesAsync(batch);
+
+mockSender
+    .Verify(sender => sender.SendMessagesAsync(
+        It.IsAny<ServiceBusMessageBatch>(),
+        It.IsAny<CancellationToken>()),
+    Times.Once);
+
+// For illustrative purposes, check that the messages in the batch match what the application expects to have
+// added.
+
+foreach (ServiceBusMessage message in backingList)
+{
+    Assert.IsTrue(sourceMessages.Contains(message));
+}
+Assert.AreEqual(backingList.Count, sourceMessages.Count);
+```
+
+## Session sending and receiving
+
+```C# Snippet:ServiceBus_MockingSessionReceiver
+// This sets up the mock ServiceBusSender.
+
+Mock<ServiceBusClient> mockClient = new();
+Mock<ServiceBusSender> mockSender = new();
+Mock<ServiceBusSessionReceiver> mockSessionReceiver = new();
+
+mockClient
+    .Setup(client => client.CreateSender(It.IsAny<string>()))
+    .Returns(mockSender.Object);
+
+// This sets up the ServiceBusClient to return the mock ServiceBusSessionProcessor when
+// AcceptNextSession is called.
+
+mockClient
+    .Setup(client => client.AcceptNextSessionAsync(
+        It.IsAny<string>(),
+        It.IsAny<ServiceBusSessionReceiverOptions>(),
+        It.IsAny<CancellationToken>()))
+    .ReturnsAsync(mockSessionReceiver.Object);
+
+ServiceBusClient client = mockClient.Object;
+
+// This sets up a set of messages to send to the session.
+
+List<ServiceBusMessage> testSessionMessages = new();
+Queue<ServiceBusReceivedMessage> messagesToReturn = new();
+
+var mockSessionId = "mySession";
+var numMessagesInSession = 5;
+
+for (var i=0; i<numMessagesInSession; i++)
+{
+    ServiceBusMessage sentMessage = new($"message{i}")
+    {
+        SessionId = mockSessionId,
+        MessageId = $"messageId{i}"
+    };
+    testSessionMessages.Add(sentMessage);
+}
+
+// This sets up the session sender to enqueue each message sent to the queue of messages to return.
```

Comment:
Consider: splitting this into two snippets:
- sending to a session
- receiving from a session (with a pre-defined set of messages)

Reply:
If so consider if it would be helpful to have a snippet with both? to illustrate a full loop: call send on message m -> save m in list -> return m upon call to receive?

Reply:
I think either approach would be okay.